utils/keeper-bench/Runner.cpp

```diff
@@ -458,8 +1248,232 @@ std::vector<std::shared_ptr<Coordination::ZooKeeper>> Runner::refreshConnections
 
 Runner::~Runner()
 {
-    queue->clearAndFinish();
+    if (queue)
+        queue->clearAndFinish();
     shutdown = true;
-    pool->wait();
-    generator->cleanup(*connections[0]);
+
+    if (pool)
+        pool->wait();
+
+    try
+    {
+        auto connection = getConnection(connection_infos[0], 0);
+        benchmark_context.cleanup(*connection);
+    }
+    catch (...)
+    {
+        DB::tryLogCurrentException("While trying to clean nodes");
+    }
+}
+
+namespace
+{
+
+void removeRecursive(Coordination::ZooKeeper & zookeeper, const std::string & path)
+{
+    namespace fs = std::filesystem;
+
+    auto promise = std::make_shared<std::promise<void>>();
+    auto future = promise->get_future();
+
+    Strings children;
+    auto list_callback = [promise, &children] (const Coordination::ListResponse & response)
+    {
+        children = response.names;
+        promise->set_value();
+    };
+    zookeeper.list(path, Coordination::ListRequestType::ALL, list_callback, nullptr);
+    future.get();
+
+    std::span children_span(children);
+    while (!children_span.empty())
+    {
+        Coordination::Requests ops;
+        for (size_t i = 0; i < 1000 && !children.empty(); ++i)
+        {
+            removeRecursive(zookeeper, fs::path(path) / children.back());
+            ops.emplace_back(zkutil::makeRemoveRequest(fs::path(path) / children_span.back(), -1));
+            children_span = children_span.subspan(0, children_span.size() - 1);
+        }
+        auto multi_promise = std::make_shared<std::promise<void>>();
+        auto multi_future = multi_promise->get_future();
+
+        auto multi_callback = [multi_promise] (const Coordination::MultiResponse &)
+        {
+            multi_promise->set_value();
+        };
+        zookeeper.multi(ops, multi_callback);
```

Comment:
Was `multiAsync` supposed to be here?

Reply:
no, this is `Coordination::ZooKeeper` and not `zkutil::ZooKeeper` which is really obvious when the classes have same name :smile: